### PR TITLE
add --all command

### PR DESCRIPTION
**Pull request by paetling**
review @theiman

### CODE REVIEW

File: docs/changelog.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # Changelog
 
 ## 0.1.6 (In Progress)
+  * NEW: Dusty now offers a --all option when running tests.  This will run all test suites associated with an app or lib.
```
Comment (thieman) [inline docs/changelog.md]: offers an

File: dusty/cli/test.py [hunk 2]
```diff
@@ -22,13 +23,13 @@
...
 from ..commands.test import (run_app_or_lib_tests, test_info_for_app_or_lib, pull_repos_and_sync_commands,
-                             ensure_valid_suite_name)
+                             ensure_valid_suite_name, run_all_app_or_lib_suites)
 
 def main(argv):
     args = docopt(__doc__, argv, options_first=True)
     if not args['<suite_name>']:
         return Payload(test_info_for_app_or_lib, args['<app_or_lib_name>'])
-    else:
+    elif not args['--all']:
```
Comment (thieman) [inline dusty/cli/test.py]: Nitpick but generally it's easier to understand branches if you avoid negation whenever possible. So this could be `elif args['--all']` and switched with the else branch.

File: dusty/cli/test.py [hunk 3]
```diff
@@ -2,13 +2,14 @@
 If args are passed, default arguments are dropped
 
 Usage:
-  test [options] <app_or_lib_name> [<suite_name>] [<args>...] [--recreate] [--no-pull]
+  test [options] <app_or_lib_name> [--all] [<suite_name>] [<args>...]
```
Comment (thieman) [inline dusty/cli/test.py]: Does docopt give us a way to say you can either provide --all or you can provide a suite_name with args?

**Comment by thieman** [comment]
Good on design

File: dusty/cli/test.py (hunk 3 above)
Comment (jsingle) [inline dusty/cli/test.py]: I believe it does

File: docs/changelog.md (hunk 1 above)
Comment (paetling) [inline docs/changelog.md]: why?  you say it dash dash all option, right?

File: dusty/cli/test.py (hunk 2 above)
Comment (paetling) [inline dusty/cli/test.py]: I agree, but was trying to follow the same patter as above.

File: dusty/cli/test.py (hunk 3 above)
Comment (paetling) [inline dusty/cli/test.py]: will look into this

File: docs/changelog.md (hunk 1 above)
Comment (thieman) [inline docs/changelog.md]: My internal monologue just sadi "offers a all option" but whatever, trivial point :stuck_out_tongue:

File: dusty/cli/test.py [hunk 4]
```diff
@@ -21,16 +22,23 @@
...
-                             ensure_valid_suite_name)
+from ..commands.test import (run_app_or_lib_tests, test_info_for_app_or_lib, pull_repos_and_sync,
+                             ensure_valid_suite_name, run_all_app_or_lib_suites)
 
 def main(argv):
     args = docopt(__doc__, argv, options_first=True)
-    if not args['<suite_name>']:
-        return Payload(test_info_for_app_or_lib, args['<app_or_lib_name>'])
-    else:
+    if args['<suite_name>'] == '--all':
```
Comment (thieman) [inline dusty/cli/test.py]: This is crazy hacky, if something has `--` in front of it it should definitely be an option, not an arg. If we're going to do it this way I think it'd be better to just make `all` a reserved suite name.

Comment (paetling) [inline dusty/cli/test.py]: Yeah this is why i wanted you to take a look.  Docopt refuses to distinguish --all option I am trying to provide.  If i put it after gcweb it will always interpret it as <suite_name>. I like the idea of reserving all.

File: dusty/commands/test.py [hunk 5]
```diff
@@ -60,15 +61,30 @@ def pull_repos_and_sync_commands(app_or_lib_name, pull_repos=False):
     make_test_command_files(app_or_lib_name, expanded_specs)
     if pull_repos:
         _update_test_repos(app_or_lib_name)
+    spec = expanded_specs.get_app_or_lib(app_or_lib_name)
+    sync_repos_by_specs([spec])
 
-def run_app_or_lib_tests(app_or_lib_name, suite_name, test_arguments, force_recreate=False):
+def run_app_or_lib_tests(app_or_lib_name, suite_name, test_arguments, should_exit=True, force_recreate=False):
```
Comment (thieman) [inline dusty/commands/test.py]: When is should_exit false?

File: dusty/commands/test.py [hunk 6]
```diff
@@ -60,15 +61,30 @@ def pull_repos_and_sync_commands(app_or_lib_name, pull_repos=False):
...
     ensure_test_image(client, app_or_lib_name, expanded_specs, force_recreate=force_recreate)
-    _run_tests_with_image(client, expanded_specs, app_or_lib_name, test_command, suite_name)
+    return _run_tests_with_image(client, expanded_specs, app_or_lib_name, test_command, suite_name, should_exit=should_exit)
+
+def run_all_app_or_lib_suites(app_or_lib_name, force_recreate=False):
+    expanded_specs = get_expanded_libs_specs()
+    spec = expanded_specs.get_app_or_lib(app_or_lib_name)
+    exit_code = 0
+    for index, suite_spec in enumerate(spec['test']['suites']):
+        args = [app_or_lib_name, suite_spec['name'], [], False]
```
Comment (thieman) [inline dusty/commands/test.py]: This is really hard to parse, I think it'd be more clear if `should_exit` and `force_recreate` were handled as kwargs. Also less brittle.

Comment (paetling) [inline dusty/commands/test.py]: fair enough.  will make that change.

File: dusty/commands/test.py [hunk 7]
```diff
@@ -148,5 +164,7 @@ def _run_tests_with_image(client, expanded_specs, app_or_lib_name, test_command,
         except Exception as e:
             log_to_client('Exception stopping service container {}: {}'.format(service_container, str(e)))
 
-    log_to_client('TESTS {}'.format('FAILED' if exit_code != 0 else 'PASSED'))
-    sys.exit(exit_code)
+    if should_exit:
+        log_to_client('TESTS {}'.format('FAILED' if exit_code != 0 else 'PASSED'))
+        sys.exit(exit_code)
```
Comment (thieman) [inline dusty/commands/test.py]: Really this logic shouldn't even be in here, the callers should do it.
